Create Feature/sprint ghcn for 6 land converters during the ioda-converter code spring

YoulongXia-NOAA opened the pull request:
## Description

*(Instructions: replace text in this and all sections with your own text)*

-[] Following ioda-converter code sprint instructions, modifying and updating six ioda-converters for land:  **imsfv3_scf2ioda.py,    smap_ssm2ioda.py, ghcn_snod2ioda.py,  owp_snow_obs.py, smos_ssm2ioda.py,  smap9km_ssm2ioda.py**
- [] Removing afwa_snod2ioda.py from land

### Issue(s) addressed

Link the issues to be closed with this PR
- fixes #842 

If this does not close an existing issue, then [be sure to add a Pipeline, Label, Estimate, Assignees, and Epic](https://jointcenterforsatellitedataassimilation-jedi-docs.readthedocs-hosted.com/en/latest/inside/practices/issues.html)

## Acceptance Criteria (Definition of Done)

ctest passed and converters are merged into develop branch

## Dependencies

No

## Impact

Please list the other repositories (if any) that this PR will require changes in (example below)

- [ ] ufo

YoulongXia-NOAA commented:
@rhoneyager @srherbener, now my converters have two problems: (a) I can set up for platform and sensor as a string rather than integer, and (b) for  smap9km_ssm2ioda.py, there are warning messages: 
 Variable MetaData/vegetationOpacity
 Warning: Variable 'MetaData/vegetationOpacity' is not listed in the YAML conventions file.
 Variable MetaData/surfaceFlag
 Warning: Variable 'MetaData/surfaceFlag' is not listed in the YAML conventions file.
Otherwise, all six converters are there.

YoulongXia-NOAA commented:
Okay, I can use
 :platform = 789 ;
:sensor = 432 ; as an example. I will modify other converters related to satellite ID and sensor ID in the same way.

CoryMartin-NOAA commented on src/land/ghcn_snod2ioda.py:
```diff
@@ -180,8 +178,7 @@ def assignValue(colrowValue, df400):
 
-        DimDict['nlocs'] = len(self.outdata[('datetime', 'MetaData')])
-        AttrData['nlocs'] = np.int32(DimDict['nlocs'])
+        DimDict['Location'] = len(self.outdata[('datetime', 'MetaData')])
```
I think these are now supposed to be `dateTime`. May require my PR to be merged in before it will fully work.

YoulongXia-NOAA replied:
Corrected.

CoryMartin-NOAA commented on src/land/imsfv3_scf2ioda.py:
```diff
@@ -136,8 +132,7 @@ def _read(self):
                     self.outdata[self.varDict[iodavar]['qcKey']] = qdflg
-        DimDict['nlocs'] = len(self.outdata[('datetime', 'MetaData')])
-        AttrData['nlocs'] = np.int32(DimDict['nlocs'])
+        DimDict['Location'] = len(self.outdata[('datetime', 'MetaData')])
```
`dateTime` again

YoulongXia-NOAA replied:
Corrected.

CoryMartin-NOAA commented on src/land/owp_snow_obs.py:
```diff
@@ -32,11 +32,11 @@
 
 # obs file name -> ioda file name
-output_var_dict = {'snow_depth_m': 'snow_depth', 'snow_water_equivalent_mm': 'swe'}
+output_var_dict = {'snow_depth_m': 'totalSnowDepth', 'snow_water_equivalent_mm': 'swe'}
```
Presumably `swe` also needs updated?

YoulongXia-NOAA replied:
swe -> snowWaterEquivalent, units is changed: mm -> kg m-2

CoryMartin-NOAA commented on src/land/owp_snow_obs.py:
```diff
@@ -108,7 +106,7 @@ def __init__(
     def _read(self):
         # print(f"Reading: {self.file_in}")
-        self.attr_data['obs_file'] = str(self.file_in.split('/')[-1])
+        #self.attr_data['obs_file'] = str(self.file_in.split('/')[-1])
```
Remove commented out line if no longer needed

CoryMartin-NOAA commented on src/land/owp_snow_obs.py:
```diff
@@ -206,9 +203,8 @@ def _read(self):
-        nlocs = len(self.data[('datetime', 'MetaData')])
-        dim_dict['nlocs'] = nlocs
-        attr_data['nlocs'] = np.int32(nlocs)
+        Location = len(self.data[('datetime', 'MetaData')])
```
`dateTime` again

CoryMartin-NOAA replied:
Also can you condense these two lines into one?
`dim_dict['Location'] = len(self.data[('dateTime', 'MetaData')])`

YoulongXia-NOAA replied:
Yes, corrected and condensed.

CoryMartin-NOAA commented on src/land/smap9km_ssm2ioda.py:
```diff
@@ -127,17 +124,16 @@ def _read(self):
         self.varAttrs[('longitude', 'MetaData')]['units'] = 'degree_east'
         self.outdata[('surfaceFlag', 'MetaData')] = sflg
-        self.varAttrs[('surfaceFlag', 'MetaData')]['units'] = 'unitless'
+        #self.varAttrs[('surfaceFlag', 'MetaData')]['units'] = 'unitless'
```
Remove these commented out lines.

YoulongXia-NOAA replied:
Removed

CoryMartin-NOAA commented on src/land/smap9km_ssm2ioda.py:
```diff
@@ -127,17 +124,16 @@ def _read(self):
 
-        DimDict['nlocs'] = len(self.outdata[('datetime', 'MetaData')])
-        AttrData['nlocs'] = np.int32(DimDict['nlocs'])
+        DimDict['Location'] = len(self.outdata[('datetime', 'MetaData')])
```
`dateTime` again

YoulongXia-NOAA replied:
Corrected

CoryMartin-NOAA commented on src/land/smos_ssm2ioda.py:
```diff
@@ -125,8 +123,7 @@ def _read(self):
 
-        DimDict['nlocs'] = len(self.outdata[('datetime', 'MetaData')])
-        AttrData['nlocs'] = np.int32(DimDict['nlocs'])
+        DimDict['Location'] = len(self.outdata[('datetime', 'MetaData')])
```
`dateTime` again

YoulongXia-NOAA replied:
corrected.

YoulongXia-NOAA commented:
@CoryMartin-NOAA, I have addressed all issues you indicated except for datetime ->dateTime. This issue will be addressed soon when other works are ready for it.

gthompsnJCSDA reviewed (changes requested):
Where Cory marked the dateTime variables, you said you fixed/corrected them, yet when I checked out the branch, they were not changed.  This probably means you did not yet do a git add/commit/push of your changes.  Please do that soon.

Also, I found a couple spelling mistakes with your spelling of ```satelliate``` that should be fixed to ```satellite```

Otherwise, if the ioda-validate.x application works after these changes, then I am good to approve the PR.

YoulongXia-NOAA commented:
> Where Cory marked the dateTime variables, you said you fixed/corrected them, yet when I checked out the branch, they were not changed. This probably means you did not yet do a git add/commit/push of your changes. Please do that soon.
> 
> Also, I found a couple spelling mistakes with your spelling of `satelliate` that should be fixed to `satellite`
> 
> Otherwise, if the ioda-validate.x application works after these changes, then I am good to approve the PR.

@gthompsnJCSDA, I changed datetime into dateTime, do make -j8 and run ctest, it aborted. I talked Cory and will wait for his work to be merged and see if I can run ctest. Two spelling of `satelliate` was fixed with `satellite`.

CoryMartin-NOAA commented:
@gthompsnJCSDA this is because @YoulongXia-NOAA is using strings, so the PR you just approved (and then the one in ioda) is needed before Youlong's converter can use `dateTime` properly.

YoulongXia-NOAA commented:
line 78-80, platform and sensor need to put integer. An example for SMAP is below:
self.satellite = 789
self.sensor = 432
AttrData["platform"] = np.array([self.satellite], dtype=np.int32)
AttrData["sensor"] = np.array([self.sensor], dtype=np.int32)

YoulongXia-NOAA commented:
> @gthompsnJCSDA this is because @YoulongXia-NOAA is using strings, so the PR you just approved (and then the one in ioda) is needed before Youlong's converter can use `dateTime` properly.

I think that now most of us are using strings. We need some instructions and guidance to do further modifications. Thank you very much for your efforts in helping us to move forward. @CoryMartin-NOAA @gthompsnJCSDA

CoryMartin-NOAA commented on src/land/smap9km_ssm2ioda.py:
```diff
@@ -115,29 +113,26 @@ def _read(self):
-        self.varAttrs[('surfaceFlag', 'MetaData')]['units'] = 'unitless'
         self.outdata[('vegetationOpacity', 'MetaData')] = vegop
-        self.varAttrs[('vegetationOpacity', 'MetaData')]['units'] = 'unitless'
+        #self.varAttrs[('vegetationOpacity', 'MetaData')]['units'] = 'unitless'
```
Remove commented out line

YoulongXia-NOAA replied:
Removed

CoryMartin-NOAA commented on src/land/imsfv3_scf2ioda.py:
```diff
@@ -71,22 +70,19 @@ def _read(self):
 
         # read netcdf file
         ncd = nc.Dataset(self.filename)
-        AttrData["sensor"] = "IMS Multisensor"
```
Is there an integer we can use instead?

YoulongXia-NOAA replied:
Removed

srherbener commented:
@gthompsnJCSDA, @CoryMartin-NOAA have your concerns been addressed? If so, please approve to indicate that. Thanks!

srherbener reviewed (approved):
Looks good (given that @CoryMartin-NOAA and @gthompsnJCSDA concerns are addressed). Thanks!

srherbener commented:
> Looks good (given that @CoryMartin-NOAA and @gthompsnJCSDA concerns are addressed). Thanks!

And the conflicts are fixed. Thanks!

gthompsnJCSDA commented:
@YoulongXia-NOAA   Did you make a corresponding pull request in the ioda:feature/sprint-ioda-converters to add the ```vegetationOpacity``` variable yet?  Also, there is already a variable called ```earthSurfaceType``` in the ```ObsSpace.yaml``` that you can use for the variable you have called ```surfaceFlag```.

YoulongXia-NOAA commented:
> @YoulongXia-NOAA Did you make a corresponding pull request in the ioda:feature/sprint-ioda-converters to add the `vegetationOpacity` variable yet? Also, there is already a variable called `earthSurfaceType` in the `ObsSpace.yaml` that you can use for the variable you have called `surfaceFlag`.

Not yet, will do on Monday. I also need to solve the conflict simultaneously. Will let you know when I am ready. Thank you.

YoulongXia-NOAA commented:
> @YoulongXia-NOAA Did you make a corresponding pull request in the ioda:feature/sprint-ioda-converters to add the `vegetationOpacity` variable yet? Also, there is already a variable called `earthSurfaceType` in the `ObsSpace.yaml` that you can use for the variable you have called `surfaceFlag`.

Not yet, will do on Monday. I also need to solve the conflict simultaneously. Will let you know when I am ready. Thank you.

YoulongXia-NOAA commented:
Yes, it is https://github.com/JCSDA-internal/ioda/pull/659 @gthompsnJCSDA

rhoneyager commented:
The ioda PR is merged. @gthompsnJCSDA, @YoulongXia-NOAA and I will look at the branch conflicts later today.

YoulongXia-NOAA commented:
@Ryan, ctest fails is due that Andy code still use datetime rather than dateTime. After I changed, ctest works.  I also fixed for station_id with stationIdentification and ioda-validation.x passed.  So we can cancel tomorrow meeting. Thank you and have nice evening.
